### ISSUE TITLE
Add Iterable.takeUntil() or similar

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon.language#739]
The method:

```ceylon
{Element*} takeUntil(Boolean stopping(Element element))
```
would be similar to `takeWhile`, with the following differences:

* The passed in function would return `true` to signal the end of the stream, rather than `false`
* The `element` that caused the "stop" would be included in the resultant stream (`takeWhile` discards the "final" element).

This is useful if your goal is to take all elements up to and including the "element you are looking for". `found` could also be used as the name for the passed in function.

**Comment by quintesse** [comment]
Although I understand why you want this I personally feel it's wrong to use `takeUntil` for this. Logically the only difference between a `takeWhile` and a `takeUntil` should be the fact that the stop condition is inverted. I think people might find it hard to remember which of the two is the one that includes and which is the one that excludes the element at the point the stop condition is met.

I'd rather have a boolean argument `inclusive` or a different name for the method itself, ie `takeWhileInclusive` (personally I prefer being verbose but clear).

**Comment by jvasileff** [comment]
Yeah, I agree that `takeUntil` does not imply inclusion of the stopper.

There's a similar discussion [here](https://github.com/ReactiveX/RxJava/issues/1649), with the following names suggested:

- takeUntil
- takeUntilInclusive
- takeWhileInclusive
- skipAfter
- terminateAfter

**Comment by tombentley** [comment]
Can this wait till 1.3?

**Comment by gavinking** [comment]
Yes. (OTOH if someone can think of a decent name for it, it could be implemented in 5 mins.) In the absence of such a name I will reassign it.

**Comment by quintesse** [comment]
The other option would be to introduce a `delay()` method that takes a first value and a function that together work to return each value one invocation "late", eg `seq.takeWhile(delay(true, (c)=>c!=','))`. Just brilliant ;)

**Comment by gavinking** [comment]
@quintesse interesting idea...

**Comment by jvasileff** [comment]
IMO `delay` is too complicated to obviate the need for the unnamed function, but very cool @quintesse!

I almost adopted it for my project, but `terminateAfter(p)` can return the nonempty `{Element+}`, and `takeWhile(delay(true, p))` can't.

**Comment by luolong** [comment]
I second that. As for names, what about these:

- `Iterable.takeWhile`, `Iterable.takeUntil`
- `Iterable.skipWhile`, `Iterable.skipUntil`

The convention would be that methodes with `*While` are inclusive and methods with `*Until` are exclusive.

(I have no idea what would be the use cases for `skipAfter` and `terminateAfter` -- to me it seems like an alias to `takeUntil`)

**Comment by quintesse** [comment]
@luolong sure, we could do that but to me that just means that people will continuously look up the docs to see what was the difference again between the two.

We had this (very long) discussion with `span()` and `measure()` and even though I've done work on the implementation of those two I *still* confuse them (and even worse when using `x..y` vs `x:y` because you won't even get documentation hints).

Now if this was a common convention in other languages and everybody was accustomed to the idea perhaps.... maybe, just maybe, we could make a comparison with while loops vs do-until loops and make it easier to remember that way. Hmmm.

**Comment by FroMage** [comment]
Relevant Perl manual:

> Lists may be assigned to only when each element of the list is itself legal to assign to:
> ```perl
> ($a, $b, $c) = (1, 2, 3);
> ($map{'red'}, $map{'blue'}, $map{'green'}) = (0x00f, 0x0f0, 0xf00);
> ```
> An exception to this is that you may assign to undef in a list. This is useful for throwing away some of the return values of a function:
> ```perl
> ($dev, $ino, undef, undef, $uid, $gid) = stat($file);
> ```
> As of Perl 5.22, you can also use (undef)x2 instead of undef, undef . (You can also do ($x) x 2 , which is less useful, because it assigns to the same variable twice, clobbering the first value assigned.)

> The final element of a list assignment may be an array or a hash:
> ```perl
>     ($a, $b, @rest) = split;
>     my($a, $b, %rest) = @_;
> ```
> You can actually put an array or hash anywhere in the list, but the first one in the list will soak up all the values, and anything after it will become undefined. This may be useful in a my() or local().

**Comment by FroMage** [comment]
Sorry, wrong thread. I deleted my comment.
